### PR TITLE
Waiting flag plumbed through provision + some extras

### PR DESCRIPTION
**Pull request by chrisnojima**
- [x] adding waiting flag for provision flow and hooking up buttons. 
- [x] adding more delay options to engine and a waiting handler so we can avoid making explicit calls to waiting on/off
- [x] cleaning up some unused provision stuff from < login3, moving helpers to where they're used
- [x] plumbed through passphrase error
- [x] made login/provision available to admins, we need design feedback now
- [x] simplified the main window resizing logic in the desktop nav

@keybase/react-hackers

### CODE REVIEW

File: desktop/README.md [hunk 1]
```diff
@@ -41,6 +41,7 @@ Environment variables:
  - `KEYBASE_SHOW_DEVTOOLS` (bool): Show devtools
  - `KEYBASE_FEATURES`: Feature flags
  - `KEYBASE_RPC_DELAY`: Number of ms to delay all RPC calls (requires debug mode)
+ - `KEYBASE_RPC_DELAY_RESULT`: Number of ms to delay all RPC call callbacks (requires debug mode)
```
Comment (chrisnojima) [inline desktop/README.md]: we want to delay the result, not the call, so we can see waiting spinners

File: shared/actions/login/index.js [hunk 2]
```diff
@@ -25,6 +26,21 @@ import {overrideLoggedInTab} from '../../local-debug'
 import type {DeviceRole} from '../../constants/login'
 import openURL from '../../util/open-url'
 
+function makeWaitingHandler (dispatch: Dispatch): {waitingHandler: (waiting: boolean) => void} {
```
Comment (chrisnojima) [inline shared/actions/login/index.js]: helper to dispatch out state changes due to waiting changing

File: shared/actions/login/index.js [hunk 3]
```diff
@@ -94,7 +112,10 @@ export function login (): AsyncAction {
 
           if (!isRPCCancelError(error)) {
             dispatch(routeAppend({
-              parseRoute: {componentAtTop: {component: Error, props: {error}}}
+              parseRoute: {componentAtTop: {component: Error, props: {
+                error,
+                onBack: () => dispatch(cancelLogin())
```
Comment (chrisnojima) [inline shared/actions/login/index.js]: onBack was missing, you'd crash if you clicked back on this page

File: shared/actions/login/index.js [hunk 4]
```diff
@@ -317,10 +343,12 @@ function askForCodePage (cb) : AsyncAction {
   }
 }
 
-function cancelLogin (response) : AsyncAction {
+function cancelLogin (response: ?responseError) : AsyncAction {
   return (dispatch, getState) => {
     dispatch(navBasedOnLoginState())
-    engine.cancelRPC(response)
+    if (response) {
```
Comment (chrisnojima) [inline shared/actions/login/index.js]: we call this when we've errored out and we don't actually have a response

File: shared/actions/login/index.js [hunk 5]
```diff
@@ -386,7 +415,8 @@ function makeKex2IncomingMap (dispatch, getState) : incomingCallMapType {
               passphrase,
               storeSecret: false
             }),
-            onBack: () => dispatch(cancelLogin(response))
+            onBack: () => dispatch(cancelLogin(response)),
+            error: retryLabel
```
Comment (chrisnojima) [inline shared/actions/login/index.js]: plumb through errors

File: shared/devices/existing-device/index.render.desktop.js [hunk 6]
```diff
@@ -1,7 +1,7 @@
 import React from 'react'
-import {Text, Icon} from '../../../common-adapters'
-import {globalStyles, globalColors, transition} from '../../../styles/style-guide'
-import Container from '../../forms/container'
+import {Text, Icon} from '../../common-adapters'
+import {globalStyles, globalColors, transition} from '../../styles/style-guide'
+import Container from '../../login/forms/container'
```
Comment (chrisnojima) [inline shared/devices/existing-device/index.render.desktop.js]: only used by device screen now

File: shared/engine/index.js [hunk 7]
```diff
@@ -33,7 +33,13 @@ class Engine {
     this.program = 'keybase.1'
     this.rpcClient = new RpcClient(
       new Transport(
-        payload => { this._rpcIncoming(payload) },
+        payload => {
+          if (__DEV__ && process.env.KEYBASE_RPC_DELAY_RESULT) {
```
Comment (chrisnojima) [inline shared/engine/index.js]: we have a new flag to delay our calls

File: shared/engine/index.js [hunk 8]
```diff
@@ -168,6 +174,9 @@ class Engine {
         }
         if (response) {
           this.sessionIDToResponse[sessionID] = null
+          if (waitingHandler) {
+            waitingHandler(true, method, sessionID)
```
Comment (chrisnojima) [inline shared/engine/index.js]: we mark us as waiting for a response or not

File: desktop/app/main-window.js [hunk 9]
```diff
@@ -10,8 +9,7 @@ export default function () {
     resolveRoot(`renderer/index.html?src=${hotPath('index.bundle.js')}`), {
       useContentSize: true,
       width: globalResizing.login.width,
-      height: globalResizing.login.height,
-      show: flags.login
```
Comment (chrisnojima) [inline desktop/app/main-window.js]: showing is handled by the nav now

File: shared/login/index.js [hunk 10]
```diff
@@ -27,9 +26,6 @@ function loginRouter (currentPath: Map<string, string>, uri: URI): any {
         break
       case 'signup':
         return signupRouter(currentPath, uri)
-      case 'register':
```
Comment (chrisnojima) [inline shared/login/index.js]: this page is gone

File: shared/nav.desktop.js [hunk 11]
```diff
@@ -64,51 +65,77 @@ class Nav extends Component {
   }
 
   _resizeWindow () {
```
Comment (chrisnojima) [inline shared/nav.desktop.js]: this thing was very fragile before and had a bunch of edge cases, this should be better

File: shared/nav.desktop.js [hunk 12]
```diff
@@ -121,6 +148,10 @@ class Nav extends Component {
       )
     }
 
+    if (!flags.mainWindow) {
```
Comment (chrisnojima) [inline shared/nav.desktop.js]: They shouldn't see this but just in case

File: shared/util/feature-flags.js [hunk 13]
```diff
@@ -7,23 +7,27 @@ import getenv from 'getenv'
...
   'login': boolean,
+  'mainWindow': boolean,
   'mobileAppsExist': boolean
 }
 
 let features = getenv.array('KEYBASE_FEATURES', 'string', '')
 
 const admin = features.includes(adminKey)
-const login = features.includes(loginKey)
+const login = features.includes(loginKey) || admin
```
Comment (chrisnojima) [inline shared/util/feature-flags.js]: let's let design see this

**Comment by chrisnojima** [comment]
@keybase/react-hackers bump

File: shared/actions/login/index.js [hunk 14]
```diff
@@ -57,6 +73,7 @@ export function navBasedOnLoginState () :AsyncAction {
 function getAccounts (): AsyncAction {
   return dispatch => {
     const params: login_getConfiguredAccounts_rpc = {
+      ...makeWaitingHandler(dispatch),
```
Comment (MarcoPolo) [inline shared/actions/login/index.js]: Isn't this just replicating what bindActionCreators does?

I think if you do
`waitingHandler: bindActionCreators(waitingForResponse, dispatch)`
it's the same thing and less of our own code

File: shared/nav.desktop.js [hunk 15]
```diff
@@ -64,51 +65,77 @@ class Nav extends Component {
...
+        // Did we actually change the size
+        if (width !== globalResizing.login.width && height !== globalResizing.login.height) {
+          this._originalSize = {width, height}
+        }
+
+        currentWindow.setContentSize(globalResizing.login.width, globalResizing.login.height, true)
+        currentWindow.setResizable(false)
+        currentWindow.show()
+      } else if (oldWasLogin) {
+        if (!flags.mainWindow) {
```
Comment (MarcoPolo) [inline shared/nav.desktop.js]: seems weird to put this in the middle of the resizing logic

File: shared/reducers/login.js [hunk 16]
```diff
@@ -45,7 +45,8 @@ type LoginState = {
     name: string,
     id: string,
     type: 'mobile' | 'laptop'
```
Comment (MarcoPolo) [inline shared/reducers/login.js]: add a todo here to use device type. `laptop` doesn't exist anywhere (I don't think)

Comment (chrisnojima) [inline shared/reducers/login.js]: turns out this isn't used so i'm killing it

File: shared/nav.desktop.js (hunk 15 above)
Comment (chrisnojima) [inline shared/nav.desktop.js]: i renamed it so its about checking the tab. where the consequences are to resize/hide

**Comment by chrisnojima** [comment]
@MarcoPolo updated

File: shared/nav.desktop.js [hunk 17]
```diff
@@ -56,59 +57,85 @@ class Nav extends Component {
...
+    if (this._checkingTab) {
       return
     }
 
-    if (this._renderedActiveTab === loginTab) {
-      const [width, height] = currentWindow.getContentSize()
+    this._checkingTab = true
+
+    setImmediate(() => {
+      this._checkingTab = false
```
Comment (MarcoPolo) [inline shared/nav.desktop.js]: Do we want to set it to false at the beginning? Wouldn't it be possible to have multiple things in here if that's the case?

**Comment by MarcoPolo** [comment]
:+1: % Q

File: shared/nav.desktop.js (hunk 17 above)
Comment (chrisnojima) [inline shared/nav.desktop.js]: this is all single threaded and nothing in here calls back into checkTabChanged. i'm just worried about some logic below not setting it and then we're stuck (like an early return, of which there are some below)

**Pull request merged**